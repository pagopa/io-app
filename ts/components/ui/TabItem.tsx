import React from "react";
import { GestureResponderEvent, Pressable, StyleSheet } from "react-native";
import Animated, {
  Extrapolate,
  interpolate,
  interpolateColor,
  useAnimatedStyle,
  useDerivedValue,
  useSharedValue,
  withSpring
} from "react-native-reanimated";
import {
  IOColors,
  IOIcons,
  Icon,
  hexToRgba,
  HSpacer,
  IOScaleValues,
  IOSpringValues
} from "@pagopa/io-app-design-system";
import { WithTestID } from "../../types/WithTestID";
import { LabelHeader } from "../core/typography/LabelHeader";
import { useSpringPressProgressValue } from "./utils/hooks/useSpringPressProgressValue";

type ColorMode = "light" | "dark";

export type TabItem = WithTestID<{
  label: string;
  color?: ColorMode;
  selected?: boolean;
  fullWidth?: boolean;
  // Icons
  icon?: IOIcons;
  iconSelected?: IOIcons;
  // Accessibility
  accessibilityLabel: string;
  accessibilityHint?: string;
  // Events
  onPress?: (event: GestureResponderEvent) => void;
}>;

type ColorStates = {
  background: {
    default: string;
    selected: string;
    pressed: string;
  };
  foreground: {
    default: IOColors;
    selected: IOColors;
  };
};

const mapColorStates: Record<NonNullable<TabItem["color"]>, ColorStates> = {
  light: {
    background: {
      default: "#ffffff00",
      selected: IOColors["grey-50"],
      pressed: IOColors["grey-50"]
    },
    foreground: {
      default: "grey-650",
      selected: "black"
    }
  },
  dark: {
    background: {
      default: "#ffffff00",
      selected: IOColors.white,
      pressed: IOColors.white
    },
    foreground: {
      default: "white",
      selected: "grey-850"
    }
  }
};

const TabItem = ({
  label,
  color = "light",
  selected = false,
  fullWidth = false,
  accessibilityLabel,
  accessibilityHint,
  testID,
  onPress,
  icon,
  iconSelected
}: TabItem) => {
  const {
    progress: progressPressed,
    onPressIn,
    onPressOut
  } = useSpringPressProgressValue(IOSpringValues.selection);

  const colors = mapColorStates[color];
  const foregroundColor = colors.foreground[selected ? "selected" : "default"];

  const opaquePressedBackgroundColor = hexToRgba(
    colors.background.pressed,
    0.1
  );

  const isSelected: Animated.SharedValue<number> = useSharedValue(0);
  const progressSelected = useDerivedValue(() =>
    withSpring(isSelected.value, IOSpringValues.selection)
  );

  React.useEffect(() => {
    // eslint-disable-next-line functional/immutable-data
    isSelected.value = selected ? 1 : 0;
  }, [isSelected, selected]);

  // Interpolate animation values from `pressed` values
  const animatedStyle = useAnimatedStyle(() => {
    // Link color states to the pressed states
    const pressedBackgroundColor = interpolateColor(
      progressPressed.value,
      [0, 1],
      [colors.background.default, opaquePressedBackgroundColor]
    );

    const selectedBackgroundColor = interpolateColor(
      progressSelected.value,
      [0, 1],
      [opaquePressedBackgroundColor, colors.background.selected]
    );

    // Scale down button slightly when pressed
    const scale = interpolate(
      progressPressed.value,
      [0, 1],
      [1, IOScaleValues?.basicButton?.pressedState],
      Extrapolate.CLAMP
    );

    return {
      backgroundColor: selected
        ? selectedBackgroundColor
        : pressedBackgroundColor,
      transform: [{ scale }]
    };
  }, [progressPressed, progressSelected, selected]);

  const activeIcon = selected ? iconSelected ?? icon : icon;

  return (
    <Pressable
      accessibilityLabel={accessibilityLabel}
      accessibilityHint={accessibilityHint}
      accessibilityRole={"button"}
      testID={testID}
      onPress={onPress}
      onPressIn={onPressIn}
      onPressOut={onPressOut}
      accessible={true}
    >
      <Animated.View
        style={[styles.container, animatedStyle, fullWidth && styles.fullWidth]}
      >
        {activeIcon && (
          <>
            <Icon name={activeIcon} color={foregroundColor} size={16} />
            <HSpacer size={4} />
          </>
        )}
        <LabelHeader color={foregroundColor}>{label}</LabelHeader>
      </Animated.View>
    </Pressable>
  );
};

const styles = StyleSheet.create({
  container: {
    flexDirection: "row",
    alignItems: "center",
    paddingHorizontal: 16,
    paddingVertical: 8,
    borderRadius: 65,
    justifyContent: "center",
    alignSelf: "flex-start"
  },
  fullWidth: {
    alignSelf: "auto"
  }
});

export { TabItem };
