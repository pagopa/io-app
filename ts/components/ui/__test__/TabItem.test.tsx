import TestRenderer from "react-test-renderer";
import React from "react";
import { TabItem } from "../TabItem";

describe("TabItem", () => {
  it("should match the snapshot with default props", () => {
    expect(
      TestRenderer.create(
        <TabItem
          label="Label tab"
          accessibilityLabel="Label tab"
          accessibilityHint="Label tab"
          onPress={() => undefined}
        />
      )
    ).toMatchSnapshot();
  });

  it("should match the snapshot with icons", () => {
    expect(
      TestRenderer.create(
        <TabItem
          label="Label tab"
          accessibilityLabel="Label tab"
          accessibilityHint="Label tab"
          onPress={() => undefined}
          icon="starEmpty"
          iconSelected="starFilled"
        />
      )
    ).toMatchSnapshot();
  });

  it("should match the snapshot with dark color", () => {
    expect(
      TestRenderer.create(
        <TabItem
          label="Label tab"
          accessibilityLabel="Label tab"
          accessibilityHint="Label tab"
          onPress={() => undefined}
          icon="starEmpty"
          iconSelected="starFilled"
          color="dark"
        />
      )
    ).toMatchSnapshot();
  });

  it("should match the snapshot with dark color and selected state", () => {
    expect(
      TestRenderer.create(
        <TabItem
          label="Label tab"
          accessibilityLabel="Label tab"
          accessibilityHint="Label tab"
          onPress={() => undefined}
          icon="starEmpty"
          iconSelected="starFilled"
          color="dark"
          selected={true}
        />
      )
    ).toMatchSnapshot();
  });

  it("should match the snapshot with full width", () => {
    expect(
      TestRenderer.create(
        <TabItem
          label="Label tab"
          accessibilityLabel="Label tab"
          accessibilityHint="Label tab"
          onPress={() => undefined}
          icon="starEmpty"
          iconSelected="starFilled"
          color="dark"
          selected={true}
          fullWidth={true}
        />
      )
    ).toMatchSnapshot();
  });
});
