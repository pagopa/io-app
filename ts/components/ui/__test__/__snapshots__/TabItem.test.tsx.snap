// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabItem should match the snapshot with dark color 1`] = `
<View
  accessibilityHint="Label tab"
  accessibilityLabel="Label tab"
  accessibilityRole="button"
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "alignSelf": "flex-start",
          "borderRadius": 65,
          "flexDirection": "row",
          "justifyContent": "center",
          "paddingHorizontal": 16,
          "paddingVertical": 8,
        },
        Object {
          "backgroundColor": undefined,
          "transform": Array [
            Object {
              "scale": undefined,
            },
          ],
        },
        false,
      ]
    }
  >
    <RNSVGSvgView
      accessibilityElementsHidden={true}
      accessibilityLabel=""
      accessible={false}
      align="xMidYMid"
      bbHeight={16}
      bbWidth={16}
      color={4294967295}
      focusable={false}
      height={16}
      importantForAccessibility="no-hide-descendants"
      meetOrSlice={0}
      minX={0}
      minY={0}
      style={
        Array [
          Object {
            "backgroundColor": "transparent",
            "borderWidth": 0,
          },
          Object {
            "color": "#FFFFFF",
          },
          Object {
            "flex": 0,
            "height": 16,
            "width": 16,
          },
        ]
      }
      tintColor={4294967295}
      vbHeight={24}
      vbWidth={24}
      width={16}
    >
      <RNSVGGroup>
        <RNSVGPath
          clipRule={0}
          d="M10.2605 2.03943c.7664-1.35075 2.7127-1.35075 3.4791 0l2.729 4.81004 5.4179 1.10907c1.5215.31145 2.1229 2.16256 1.0751 3.30876l-3.7313 4.0819.6194 5.4954c.174 1.5433-1.4006 2.6873-2.8146 2.045L12 20.6022l-5.03505 2.2874c-1.41396.6423-2.98857-.5017-2.81462-2.045l.61945-5.4954-3.73131-4.0819C-.00935 10.1211.59211 8.26998 2.11356 7.95854l5.41793-1.10907 2.72901-4.81004ZM12 3.02637 9.27102 7.83641a2.00003 2.00003 0 0 1-1.33844.97243L2.51465 9.91791l3.7313 4.08179c.38952.4262.57591.9998.51124 1.5735l-.61944 5.4955 5.03505-2.2874c.5257-.2387 1.1288-.2388 1.6544 0l5.0351 2.2874-.6194-5.4955a2.00015 2.00015 0 0 1 .5112-1.5735l3.7313-4.08179-5.4179-1.10907a1.99973 1.99973 0 0 1-1.3384-.97243L12 3.02637Z"
          fill={
            Array [
              2,
            ]
          }
          fillRule={0}
          propList={
            Array [
              "fill",
              "fillRule",
            ]
          }
        />
      </RNSVGGroup>
    </RNSVGSvgView>
    <View
      style={
        Object {
          "width": 4,
        }
      }
    />
    <Text
      color="white"
      defaultColor="black"
      defaultWeight="Regular"
      font="ReadexPro"
      fontStyle={
        Object {
          "fontSize": 14,
          "lineHeight": 18,
        }
      }
      style={
        Array [
          Object {
            "fontSize": 14,
            "lineHeight": 18,
          },
          Object {
            "color": "#FFFFFF",
            "fontFamily": "Readex Pro",
            "fontStyle": "normal",
            "fontWeight": "400",
          },
        ]
      }
      weight="Regular"
    >
      Label tab
    </Text>
  </View>
</View>
`;

exports[`TabItem should match the snapshot with dark color and selected state 1`] = `
<View
  accessibilityHint="Label tab"
  accessibilityLabel="Label tab"
  accessibilityRole="button"
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "alignSelf": "flex-start",
          "borderRadius": 65,
          "flexDirection": "row",
          "justifyContent": "center",
          "paddingHorizontal": 16,
          "paddingVertical": 8,
        },
        Object {
          "backgroundColor": undefined,
          "transform": Array [
            Object {
              "scale": undefined,
            },
          ],
        },
        false,
      ]
    }
  >
    <RNSVGSvgView
      accessibilityElementsHidden={true}
      accessibilityLabel=""
      accessible={false}
      align="xMidYMid"
      bbHeight={16}
      bbWidth={16}
      color={4281019960}
      focusable={false}
      height={16}
      importantForAccessibility="no-hide-descendants"
      meetOrSlice={0}
      minX={0}
      minY={0}
      style={
        Array [
          Object {
            "backgroundColor": "transparent",
            "borderWidth": 0,
          },
          Object {
            "color": "#2B2E38",
          },
          Object {
            "flex": 0,
            "height": 16,
            "width": 16,
          },
        ]
      }
      tintColor={4281019960}
      vbHeight={24}
      vbWidth={24}
      width={16}
    >
      <RNSVGGroup>
        <RNSVGPath
          clipRule={0}
          d="M10.2576 2.04309c.7663-1.35075 2.7127-1.35075 3.479 0l2.7291 4.81004 5.4179 1.10907c1.5215.31145 2.1229 2.1625 1.0751 3.3088l-3.7313 4.0818.6194 5.4955c.174 1.5432-1.4006 2.6873-2.8146 2.0449l-5.0351-2.2873-5.03508 2.2873c-1.41396.6424-2.98857-.5017-2.81462-2.0449l.61945-5.4955-3.73131-4.0818C-.01228 10.1247.58918 8.27365 2.11063 7.9622l5.41793-1.10907 2.72904-4.81004Z"
          fill={
            Array [
              2,
            ]
          }
          fillRule={0}
          propList={
            Array [
              "fill",
              "fillRule",
            ]
          }
        />
      </RNSVGGroup>
    </RNSVGSvgView>
    <View
      style={
        Object {
          "width": 4,
        }
      }
    />
    <Text
      color="grey-850"
      defaultColor="black"
      defaultWeight="Regular"
      font="ReadexPro"
      fontStyle={
        Object {
          "fontSize": 14,
          "lineHeight": 18,
        }
      }
      style={
        Array [
          Object {
            "fontSize": 14,
            "lineHeight": 18,
          },
          Object {
            "color": "#2B2E38",
            "fontFamily": "Readex Pro",
            "fontStyle": "normal",
            "fontWeight": "400",
          },
        ]
      }
      weight="Regular"
    >
      Label tab
    </Text>
  </View>
</View>
`;

exports[`TabItem should match the snapshot with default props 1`] = `
<View
  accessibilityHint="Label tab"
  accessibilityLabel="Label tab"
  accessibilityRole="button"
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "alignSelf": "flex-start",
          "borderRadius": 65,
          "flexDirection": "row",
          "justifyContent": "center",
          "paddingHorizontal": 16,
          "paddingVertical": 8,
        },
        Object {
          "backgroundColor": undefined,
          "transform": Array [
            Object {
              "scale": undefined,
            },
          ],
        },
        false,
      ]
    }
  >
    <Text
      color="grey-650"
      defaultColor="black"
      defaultWeight="Regular"
      font="ReadexPro"
      fontStyle={
        Object {
          "fontSize": 14,
          "lineHeight": 18,
        }
      }
      style={
        Array [
          Object {
            "fontSize": 14,
            "lineHeight": 18,
          },
          Object {
            "color": "#636B82",
            "fontFamily": "Readex Pro",
            "fontStyle": "normal",
            "fontWeight": "400",
          },
        ]
      }
      weight="Regular"
    >
      Label tab
    </Text>
  </View>
</View>
`;

exports[`TabItem should match the snapshot with full width 1`] = `
<View
  accessibilityHint="Label tab"
  accessibilityLabel="Label tab"
  accessibilityRole="button"
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "alignSelf": "flex-start",
          "borderRadius": 65,
          "flexDirection": "row",
          "justifyContent": "center",
          "paddingHorizontal": 16,
          "paddingVertical": 8,
        },
        Object {
          "backgroundColor": undefined,
          "transform": Array [
            Object {
              "scale": undefined,
            },
          ],
        },
        Object {
          "alignSelf": "auto",
        },
      ]
    }
  >
    <RNSVGSvgView
      accessibilityElementsHidden={true}
      accessibilityLabel=""
      accessible={false}
      align="xMidYMid"
      bbHeight={16}
      bbWidth={16}
      color={4281019960}
      focusable={false}
      height={16}
      importantForAccessibility="no-hide-descendants"
      meetOrSlice={0}
      minX={0}
      minY={0}
      style={
        Array [
          Object {
            "backgroundColor": "transparent",
            "borderWidth": 0,
          },
          Object {
            "color": "#2B2E38",
          },
          Object {
            "flex": 0,
            "height": 16,
            "width": 16,
          },
        ]
      }
      tintColor={4281019960}
      vbHeight={24}
      vbWidth={24}
      width={16}
    >
      <RNSVGGroup>
        <RNSVGPath
          clipRule={0}
          d="M10.2576 2.04309c.7663-1.35075 2.7127-1.35075 3.479 0l2.7291 4.81004 5.4179 1.10907c1.5215.31145 2.1229 2.1625 1.0751 3.3088l-3.7313 4.0818.6194 5.4955c.174 1.5432-1.4006 2.6873-2.8146 2.0449l-5.0351-2.2873-5.03508 2.2873c-1.41396.6424-2.98857-.5017-2.81462-2.0449l.61945-5.4955-3.73131-4.0818C-.01228 10.1247.58918 8.27365 2.11063 7.9622l5.41793-1.10907 2.72904-4.81004Z"
          fill={
            Array [
              2,
            ]
          }
          fillRule={0}
          propList={
            Array [
              "fill",
              "fillRule",
            ]
          }
        />
      </RNSVGGroup>
    </RNSVGSvgView>
    <View
      style={
        Object {
          "width": 4,
        }
      }
    />
    <Text
      color="grey-850"
      defaultColor="black"
      defaultWeight="Regular"
      font="ReadexPro"
      fontStyle={
        Object {
          "fontSize": 14,
          "lineHeight": 18,
        }
      }
      style={
        Array [
          Object {
            "fontSize": 14,
            "lineHeight": 18,
          },
          Object {
            "color": "#2B2E38",
            "fontFamily": "Readex Pro",
            "fontStyle": "normal",
            "fontWeight": "400",
          },
        ]
      }
      weight="Regular"
    >
      Label tab
    </Text>
  </View>
</View>
`;

exports[`TabItem should match the snapshot with icons 1`] = `
<View
  accessibilityHint="Label tab"
  accessibilityLabel="Label tab"
  accessibilityRole="button"
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "alignSelf": "flex-start",
          "borderRadius": 65,
          "flexDirection": "row",
          "justifyContent": "center",
          "paddingHorizontal": 16,
          "paddingVertical": 8,
        },
        Object {
          "backgroundColor": undefined,
          "transform": Array [
            Object {
              "scale": undefined,
            },
          ],
        },
        false,
      ]
    }
  >
    <RNSVGSvgView
      accessibilityElementsHidden={true}
      accessibilityLabel=""
      accessible={false}
      align="xMidYMid"
      bbHeight={16}
      bbWidth={16}
      color={4284705666}
      focusable={false}
      height={16}
      importantForAccessibility="no-hide-descendants"
      meetOrSlice={0}
      minX={0}
      minY={0}
      style={
        Array [
          Object {
            "backgroundColor": "transparent",
            "borderWidth": 0,
          },
          Object {
            "color": "#636B82",
          },
          Object {
            "flex": 0,
            "height": 16,
            "width": 16,
          },
        ]
      }
      tintColor={4284705666}
      vbHeight={24}
      vbWidth={24}
      width={16}
    >
      <RNSVGGroup>
        <RNSVGPath
          clipRule={0}
          d="M10.2605 2.03943c.7664-1.35075 2.7127-1.35075 3.4791 0l2.729 4.81004 5.4179 1.10907c1.5215.31145 2.1229 2.16256 1.0751 3.30876l-3.7313 4.0819.6194 5.4954c.174 1.5433-1.4006 2.6873-2.8146 2.045L12 20.6022l-5.03505 2.2874c-1.41396.6423-2.98857-.5017-2.81462-2.045l.61945-5.4954-3.73131-4.0819C-.00935 10.1211.59211 8.26998 2.11356 7.95854l5.41793-1.10907 2.72901-4.81004ZM12 3.02637 9.27102 7.83641a2.00003 2.00003 0 0 1-1.33844.97243L2.51465 9.91791l3.7313 4.08179c.38952.4262.57591.9998.51124 1.5735l-.61944 5.4955 5.03505-2.2874c.5257-.2387 1.1288-.2388 1.6544 0l5.0351 2.2874-.6194-5.4955a2.00015 2.00015 0 0 1 .5112-1.5735l3.7313-4.08179-5.4179-1.10907a1.99973 1.99973 0 0 1-1.3384-.97243L12 3.02637Z"
          fill={
            Array [
              2,
            ]
          }
          fillRule={0}
          propList={
            Array [
              "fill",
              "fillRule",
            ]
          }
        />
      </RNSVGGroup>
    </RNSVGSvgView>
    <View
      style={
        Object {
          "width": 4,
        }
      }
    />
    <Text
      color="grey-650"
      defaultColor="black"
      defaultWeight="Regular"
      font="ReadexPro"
      fontStyle={
        Object {
          "fontSize": 14,
          "lineHeight": 18,
        }
      }
      style={
        Array [
          Object {
            "fontSize": 14,
            "lineHeight": 18,
          },
          Object {
            "color": "#636B82",
            "fontFamily": "Readex Pro",
            "fontStyle": "normal",
            "fontWeight": "400",
          },
        ]
      }
      weight="Regular"
    >
      Label tab
    </Text>
  </View>
</View>
`;
